File: python/sde.py
# Copyright 2018 Wenbin He. All rights reserved.
# Use of this source code is governed by a MIT-style license that can be
# found in the LICENSE file.

# This code includes a serial python implementation of our SDE computation
# method. The use of this python implementation is to help users better
# understand the details of the SDE computation method rather than to compute
# SDE efficiently. To compute SDE for large number of surfaces on high
# resolution grids, please use our parallel implementation based on
# opengl/webgl.

from __future__ import print_function

import math
import numpy as np
from scipy import random, linalg

from tqdm import tqdm

# constants and utility functions ----------------------------------------------
kSqrt2 = math.sqrt(2.)
kSqrt2Recip = 1. / kSqrt2
kSqrt2PiRecip = 1. / math.sqrt(2. * math.pi)

def Sub(a, b):
  return [a[0] - b[0], a[1] - b[1], a[2] - b[2]]

def MulScalar(a, s):
  return [a[0] * s, a[1] * s, a[2] * s]

def MulMat(a, m):
  return [a[0] * m[0] + a[1] * m[1] + a[2] * m[2],
          a[0] * m[3] + a[1] * m[4] + a[2] * m[5],
          a[0] * m[6] + a[1] * m[7] + a[2] * m[8]]

def Dot(a, b):
  return a[0] * b[0] + a[1] * b[1] + a[2] * b[2]

def Cross(a, b):
  return [a[1] * b[2] - a[2] * b[1],
          a[2] * b[0] - a[0] * b[2],
          a[0] * b[1] - a[1] * b[0]]

def Length(a):
  return math.sqrt(a[0] * a[0] + a[1] * a[1] + a[2] * a[2])

def Normalize(a):
  l_recip = 1. / Length(a)
  return [a[0] * l_recip, a[1] * l_recip, a[2] * l_recip]

def TriangleArea(a, b, c):
  ab, ac = Sub(b, a), Sub(c, a)
  return .5 * Length(Cross(ab, ac))

def PlaneOfTriangle(a, b, c):
  ab, ac = Sub(b, a), Sub(c, a)
  n = Normalize(Cross(ab, ac))

  d = Dot(n, a)
  return n, d

def ProjOnPlane(v, n, d):
  return Sub(v, MulScalar(n, d))

def DistanceToPlane(v, n, d):
  return Dot(v, n) + d

def Sign(ax, ay, bx, by, cx, cy):
  return (ax - cx) * (by - cy) - (bx - cx) * (ay - cy)

def PointInTriangle(x, y, ax, ay, bx, by, cx, cy):
  b1 = Sign(x, y, ax, ay, bx, by) < 0.
  b2 = Sign(x, y, bx, by, cx, cy) < 0.
  b3 = Sign(x, y, cx, cy, ax, ay) < 0.
  return ((b1 == b2) and (b2 == b3))

def NormPdf(x):
  return kSqrt2PiRecip * math.exp(-.5 * x * x)

def NormCdf(x):
  return .5 * (1. + math.erf(x * kSqrt2Recip))

# Gammaj = 40
# muli = [1]
# for i in range(1, Gammaj):
#   muli.append(muli[i - 1] * i)

# def Gamma(h, a):
#   h = abs(h)
#   if (h > 4.76):
#     return 0.

#   if (a < 0.):
#     return -Gamma(h, -a)
#   if (a > 1.):
#     return .5 * NormCdf(h) + .5 * NormCdf(a * h) - \
#            NormCdf(h) * NormCdf(a * h) - Gamma(a * h, 1. / a)

#   res = math.atan(a) * .5 / math.pi
#   for j in range(Gammaj):
#     tmp = 0.
#     for i in range(j + 1):
#       tmp += pow(h, 2 * i) / pow(2., i) / muli[i]

#     res -= .5 / math.pi * pow(a, 2 * j + 1) * pow(-1., j) / (2. * j + 1.) * \
#            (1. - math.exp(-.5 * h * h) * tmp)
#   return res

gamma_table = np.fromfile("../gamma_table.raw", dtype="float32")
gamma_table = gamma_table.reshape((477, 101)).tolist()

# import matplotlib.pyplot as plt

# fig, ax = plt.subplots()
# im = ax.imshow(gamma_table)
# ax.invert_yaxis()
# fig.colorbar(im, ax=ax)
# plt.show()

def Gamma(h, a):
  h = abs(h)
  if (h > 4.76):
    return 0.

  if (a < 0.):
    return -Gamma(h, -a)
  if (a > 1.):
    return .5 * NormCdf(h) + .5 * NormCdf(a * h) - \
           NormCdf(h) * NormCdf(a * h) - Gamma(a * h, 1. / a)

  x, y = h * 100., a * 100.
  i, j = int(x), int(y)
  i, j = min(i, 475), min(j, 99)

  tx, ty = x - i, y - j
  g0 = tx * gamma_table[i + 1][j] + (1. - tx) * gamma_table[i][j]
  g1 = tx * gamma_table[i + 1][j + 1] + (1. - tx) * gamma_table[i][j + 1]

  return ty * g1 + (1. - ty) * g0
# ------------------------------------------------------------------------------

# input surfaces ---------------------------------------------------------------
verts = [[-.8, -.3,  .2],
         [ .7, -.9,  .4],
         [ .1,  .8, -.2]]
# ------------------------------------------------------------------------------

# parameters -------------------------------------------------------------------
xmin, ymin, zmin = -1., -1., -1.
xmax, ymax, zmax = 1., 1., 1.  # Physical domain that density estimation is performed on.
xdim, ydim, zdim = 65, 65, 65  # Grid resolution of the physical domain.

H = np.zeros((3, 3))  # bandwidth matrix
H[0, 0], H[0, 1], H[0, 2] = .0025,  0.,     0.
H[1, 0], H[1, 1], H[1, 2] = 0.,     .0025,  0.
H[2, 0], H[2, 1], H[2, 2] = 0.,     0.,     .0025
# ------------------------------------------------------------------------------

# SDE computation --------------------------------------------------------------
a, b, c = verts[0], verts[1], verts[2]  # three vertices of the triangle
area = TriangleArea(a, b, c)  # area of the triangle

Hi = linalg.inv(H)  # inverse of the bandwidth matrix
Hi_sqrt = linalg.sqrtm(Hi)  # square root of Hi (i.e., Hi_sqrt x Hi_sqrt = Hi)
Hi_sqrt_det = float(linalg.det(Hi_sqrt))  # determinant of Hi_sqrt
Hi_sqrt = Hi_sqrt.flatten().tolist()

# Create a numpy array to store SDE.
sde = np.zeros([zdim, ydim, xdim])

# Compute SDE for each grid point.
for i in tqdm(range(xdim * ydim * zdim)):
  # physical position of the grid point
  gk = math.floor(i / (xdim * ydim))
  gj = math.floor((i - gk * xdim * ydim) / xdim)
  gi = i % xdim

  x = [gi * (xmax - xmin) / (xdim - 1.) + xmin,
       gj * (ymax - ymin) / (ydim - 1.) + ymin,
       gk * (zmax - zmin) / (zdim - 1.) + zmin]

  # Transform the triangle based on the grid point position and the bandwidth matrix.
  ax, bx, cx = Sub(x, a), Sub(x, b), Sub(x, c)

  a_prime = MulMat(ax, Hi_sqrt)
  b_prime = MulMat(bx, Hi_sqrt)
  c_prime = MulMat(cx, Hi_sqrt)

  area_prime = TriangleArea(a_prime, b_prime, c_prime)

  ct = area / area_prime * Hi_sqrt_det

  # Map the triangle into 2D.
  n, d = PlaneOfTriangle(a_prime, b_prime, c_prime)
  a_proj = ProjOnPlane(a_prime, n, d)
  b_proj = ProjOnPlane(b_prime, n, d)
  c_proj = ProjOnPlane(c_prime, n, d)

  uaxis = Normalize(Sub(a_proj, b_proj))
  vaxis = Cross(uaxis, n)

  au = Dot(a_proj, uaxis)
  av = Dot(a_proj, vaxis)

  bu = Dot(b_proj, uaxis)
  bv = Dot(b_proj, vaxis)

  cu = Dot(c_proj, uaxis)
  cv = Dot(c_proj, vaxis)

  # Compute SDE based on bivariate normal integral.
  alpha = [0., 0., 0.]  # Used to store the bivariate normal integral over
                        # the area defined by each edge of the triangle.
  if (au != 0. or av != 0.) and \
     (bu != 0. or bv != 0.) and \
     (cu != 0. or cv != 0.):
    # segment bc
    bcu, bcv = cu - bu, cv - bv
    bxc = bu * cv - bv * cu
    h = bxc / math.sqrt(bcu * bcu + bcv * bcv)
    if (bxc != 0.):
      a1 = (bu * bcu + bv * bcv) / bxc
      a2 = (cu * bcu + cv * bcv) / bxc
      alpha[0] = abs(Gamma(h, a1) - Gamma(h, a2))
    else:
      alpha[0] = .5 if bu * cu + bv * cv < 0. else 0.

    # segment ca
    cau, cav = au - cu, av - cv
    cxa = cu * av - cv * au
    h = cxa / math.sqrt(cau * cau + cav * cav)
    if (cxa != 0.):
      a1 = (cu * cau + cv * cav) / cxa
      a2 = (au * cau + av * cav) / cxa
      alpha[1] = abs(Gamma(h, a1) - Gamma(h, a2))
    else:
      alpha[1] = .5 if cu * au + cv * av < 0. else 0.

    # segment ab
    abu, abv = bu - au, bv - av
    axb = au * bv - av * bu
    h = axb / math.sqrt(abu * abu + abv * abv)
    if (axb != 0.):
      a1 = (au * abu + av * abv) / axb
      a2 = (bu * abu + bv * abv) / axb
      alpha[2] = abs(Gamma(h, a1) - Gamma(h, a2))
    else:
      alpha[2] = .5 if au * bu + av * bv < 0. else 0.

    # Combine alpha bc, ca, and ab.
    if (PointInTriangle(0., 0., au, av, bu, bv, cu, cv)):
      sde[gk, gj, gi] = (1. - alpha[0] - alpha[1] - alpha[2]) * \
                        NormPdf(d) * ct
    else:
      coss = [0., 0., 0.]
      la = math.sqrt(au * au + av * av)
      lb = math.sqrt(bu * bu + bv * bv)
      lc = math.sqrt(cu * cu + cv * cv)
      aun, avn = au / la, av / la
      bun, bvn = bu / lb, bv / lb
      cun, cvn = cu / lc, cv / lc
      coss[0] = bun * cun + bvn * cvn
      coss[1] = cun * aun + cvn * avn
      coss[2] = aun * bun + avn * bvn

      minid = coss.index(min(coss))

      sde[gk, gj, gi] = abs(alpha[0] + alpha[1] + alpha[2] - 2. * alpha[minid]) * \
                        NormPdf(d) * ct

  # Handle spacial cases.
  elif au == 0. and av == 0.:
    # segment bc
    bcu, bcv = cu - bu, cv - bv
    bxc = bu * cv - bv * cu
    h = bxc / math.sqrt(bcu * bcu + bcv * bcv)
    if (bxc != 0.):
      a1 = (bu * bcu + bv * bcv) / bxc
      a2 = (cu * bcu + cv * bcv) / bxc
      alpha[0] = abs(Gamma(h, a1) - Gamma(h, a2))
    else:
      alpha[0] = .5 if bu * cu + bv * cv < 0. else 0.

    lb = math.sqrt(bu * bu + bv * bv)
    lc = math.sqrt(cu * cu + cv * cv)
    bun, bvn = bu / lb, bv / lb
    cun, cvn = cu / lc, cv / lc

    sde[gk, gj, gi] = (math.acos(bun * cun + bvn * cvn) / \
                       math.pi * .5 - alpha[0]) * NormPdf(d) * ct

  elif bu == 0. and bv == 0.:
    # segment ca
    cau, cav = au - cu, av - cv
    cxa = cu * av - cv * au
    h = cxa / math.sqrt(cau * cau + cav * cav)
    if (cxa != 0.):
      a1 = (cu * cau + cv * cav) / cxa
      a2 = (au * cau + av * cav) / cxa
      alpha[1] = abs(Gamma(h, a1) - Gamma(h, a2))
    else:
      alpha[1] = .5 if cu * au + cv * av < 0. else 0.

    lc = math.sqrt(cu * cu + cv * cv)
    la = math.sqrt(au * au + av * av)
    cun, cvn = cu / lc, cv / lc
    aun, avn = au / la, av / la

    sde[gk, gj, gi] = (math.acos(cun * aun + cvn * avn) / \
                       math.pi * .5 - alpha[1]) * NormPdf(d) * ct

  elif cu == 0. and cv == 0.:
    # segment ab
    abu, abv = bu - au, bv - av
    axb = au * bv - av * bu
    h = axb / math.sqrt(abu * abu + abv * abv)
    if (axb != 0.):
      a1 = (au * abu + av * abv) / axb
      a2 = (bu * abu + bv * abv) / axb
      alpha[2] = abs(Gamma(h, a1) - Gamma(h, a2))
    else:
      alpha[2] = .5 if au * bu + av * bv < 0. else 0.

    la = math.sqrt(au * au + av * av)
    lb = math.sqrt(bu * bu + bv * bv)
    aun, avn = au / la, av / la
    bun, bvn = bu / lb, bv / lb

    sde[gk, gj, gi] = (math.acos(aun * bun + avn * bvn) / \
                       math.pi * .5 - alpha[2]) * NormPdf(d) * ct

# Normalize SDE by the area of the surface.
sde /= area
# ------------------------------------------------------------------------------

# SDE visualization ------------------------------------------------------------
import vtk

# Normalize SDE for visualization.
sde /= np.max(sde)

# Create the renderer, render window, and interactor.
ren = vtk.vtkRenderer()

renWin = vtk.vtkRenderWindow()
renWin.AddRenderer(ren)

iren = vtk.vtkRenderWindowInteractor()
iren.SetRenderWindow(renWin)

# Copy SDE into a vtkImage.
reader = vtk.vtkImageImport()
reader.CopyImportVoidPointer(sde, sde.nbytes)
reader.SetDataScalarTypeToDouble()
reader.SetDataExtent(0, xdim - 1, 0, ydim - 1, 0, zdim - 1)
reader.SetWholeExtent(0, xdim - 1, 0, ydim - 1, 0, zdim - 1)
# reader.SetDataOrigin(xmin, ymin, zmin)
# reader.SetDataSpacing((xmax - xmin) / (xdim - 1.),
#                       (ymax - ymin) / (ydim - 1.),
#                       (zmax - zmin) / (zdim - 1.))

# The volume will be displayed by ray-cast alpha compositing.
volumeMapper = vtk.vtkGPUVolumeRayCastMapper()
volumeMapper.SetInputConnection(reader.GetOutputPort())
volumeMapper.SetBlendModeToComposite()

# Create transfer function mapping scalar value to opacity.
opacityTransferFunction = vtk.vtkPiecewiseFunction()
opacityTransferFunction.AddPoint(0., 0.)
opacityTransferFunction.AddPoint(1., 1.)

# Create transfer function mapping scalar value to color.
colorTransferFunction = vtk.vtkColorTransferFunction()
colorTransferFunction.AddRGBPoint(0.,    .188235, .164706, .470588)
colorTransferFunction.AddRGBPoint(.0625, .145098, .192157, .490196)
colorTransferFunction.AddRGBPoint(.125,  .133333, .227451, .501961)
colorTransferFunction.AddRGBPoint(.1875, .121569, .258824, .509804)
colorTransferFunction.AddRGBPoint(.25,   .117647, .321569, .521569)
colorTransferFunction.AddRGBPoint(.3125, .12549,  .396078, .529412)
colorTransferFunction.AddRGBPoint(.375,  .133333, .458824, .541176)
colorTransferFunction.AddRGBPoint(.4375, .141176, .517647, .54902 )
colorTransferFunction.AddRGBPoint(.5,    .172549, .568627, .537255)
colorTransferFunction.AddRGBPoint(.5625, .203922, .6,      .486275)
colorTransferFunction.AddRGBPoint(.625,  .231373, .639216, .435294)
colorTransferFunction.AddRGBPoint(.6875, .278431, .658824, .392157)
colorTransferFunction.AddRGBPoint(.75,   .337255, .721569, .337255)
colorTransferFunction.AddRGBPoint(.8125, .513725, .768627, .384314)
colorTransferFunction.AddRGBPoint(.875,  .690196, .819608, .435294)
colorTransferFunction.AddRGBPoint(.9375, .847059, .858824, .482353)
colorTransferFunction.AddRGBPoint(1.,    .901961, .878431, .556863)

# The VolumeProperty attaches the color and opacity functions to the
# volume, and sets other volume properties.
volumeProperty = vtk.vtkVolumeProperty()
volumeProperty.SetColor(colorTransferFunction)
volumeProperty.SetScalarOpacity(opacityTransferFunction)
volumeProperty.SetInterpolationTypeToLinear()

# The vtkVolume is a vtkProp3D (like a vtkActor) and controls the position
# and orientation of the volume in world coordinates.
volume = vtk.vtkVolume()
volume.SetMapper(volumeMapper)
volume.SetProperty(volumeProperty)

# Outline of the volume.
outline = vtk.vtkOutlineFilter()
outline.SetInputConnection(reader.GetOutputPort())
outlineMapper = vtk.vtkPolyDataMapper()
outlineMapper.SetInputConnection(outline.GetOutputPort())
outlineActor = vtk.vtkActor()
outlineActor.SetMapper(outlineMapper)
outlineActor.GetProperty().SetColor(0., 0., 0.)
outlineActor.GetProperty().SetLineWidth(2.)

ren.AddActor(outlineActor)
ren.AddVolume(volume)
ren.SetBackground(.94, .94, .94)
ren.GetActiveCamera().Azimuth(45.)
ren.GetActiveCamera().Elevation(30.)
ren.ResetCameraClippingRange()
ren.ResetCamera()

renWin.SetSize(600, 600)
renWin.Render()
renWin.SetWindowName("SDE of a triangle")

iren.Start()
# ------------------------------------------------------------------------------

File: javascript/sde.js
// Copyright 2018 Wenbin He. All rights reserved.
// Use of this source code is governed by a MIT-style license that can be
// found in the LICENSE file.

function SDEstimator(verts) {
  // static variables and functions --------------------------------------------
  // canvas and gl
  SDEstimator.canvas = document.createElement("canvas");

  SDEstimator.gl = SDEstimator.canvas.getContext("webgl2");
  if (!SDEstimator.gl) {
    alert("WebGL2 is not supported by this browser.");
    return;
  }

  SDEstimator.ext = SDEstimator.gl.getExtension("EXT_color_buffer_float");
  if (!SDEstimator.ext) {
    alert("EXT_color_buffer_float is not supported by this browser.");
    return;
  }
  // ---------------------------------------------------------------------------

  // public variables and functions --------------------------------------------
  // set surfaces
  this.nverts = 0;
  this.area = 0.;
  this.verts_textures = [];

  // local references of canvas and gl
  var canvas_ = SDEstimator.canvas,
      gl_ = SDEstimator.gl;

  this.SetSurfaces = function(verts) {
    // reset surfaces
    this.nverts = 0;
    this.area = 0.;
    for (var i = 0, il = this.verts_textures.length; i < il; ++i)
      gl_.deleteTexture(this.verts_textures[i]);
    this.verts_textures.length = 0;

    // Only keep triangles whose areas are greater than zero.
    var valid_verts = [];
    if (verts !== undefined) {
      for (var i = 0, il = verts.length; i < il; i += 3) {
        var area = TriangleArea(verts[i], verts[i + 1], verts[i + 2]);
        if (area > 0.) {
          valid_verts.push(verts[i].slice());
          valid_verts.push(verts[i + 1].slice());
          valid_verts.push(verts[i + 2].slice());
          this.area += area;
        }
      }
    }

    this.nverts = valid_verts.length;
    console.log("Number of valid triangles: " + this.nverts / 3);
    console.log("Total area: " + this.area);

    // Copy the surfaces into textures.
    var max_texture_size = gl_.getParameter(gl_.MAX_TEXTURE_SIZE);
    // console.log("WebGL2 maximum texture size: " +
    //             max_texture_size + " x " + max_texture_size);
    while (valid_verts.length > 0) {
      var curr_verts = valid_verts.splice(0, Math.min(valid_verts.length,
                                                      max_texture_size * 3));

      var texture = gl_.createTexture();
      gl_.bindTexture(gl_.TEXTURE_2D, texture);
      gl_.texStorage2D(gl_.TEXTURE_2D, 1, gl_.RGB32F, 3, curr_verts.length / 3);

      gl_.texParameteri(gl_.TEXTURE_2D, gl_.TEXTURE_MIN_FILTER, gl_.NEAREST);
      gl_.texParameteri(gl_.TEXTURE_2D, gl_.TEXTURE_MAG_FILTER, gl_.NEAREST);
      gl_.texParameteri(gl_.TEXTURE_2D, gl_.TEXTURE_WRAP_S, gl_.CLAMP_TO_EDGE);
      gl_.texParameteri(gl_.TEXTURE_2D, gl_.TEXTURE_WRAP_T, gl_.CLAMP_TO_EDGE);

      gl_.texSubImage2D(gl_.TEXTURE_2D, 0, 0, 0,
                        3, curr_verts.length / 3,
                        gl_.RGB, gl_.FLOAT,
                        new Float32Array(curr_verts.flat()));

      gl_.bindTexture(gl_.TEXTURE_2D, null);

      this.verts_textures.push(texture);
    }
  }

  this.SetSurfaces(verts);

  // SDE computation
  this.Compute = function(xmin, ymin, zmin,
                          xmax, ymax, zmax,  // Physical domain that density estimation is performed on.
                          xdim, ydim, zdim,  // Grid resolution of the physical domain.
                          H) {  // bandwidth matrix
    var Hi = MatInv(H);  // inverse of the bandwidth matrix
    if (Hi === undefined) {
      alert("The input bandwidth matrix is not invertible.");
      return;
    }

    var Hi_sqrt = MatSqrt(Hi);  // square root of Hi (i.e., Hi_sqrt x Hi_sqrt = Hi)
    if (Hi_sqrt === undefined) {
      alert("The input bandwidth matrix is not positive semidefinite.");
      return;
    }
    Hi_sqrt = new Float32Array(Hi_sqrt.flat());
  };
  // ---------------------------------------------------------------------------

  // private variables and functions -------------------------------------------
  // constants
  var kEps = 1e-10;
  var kMaxSweeps = 64;

  // subtraction of vectors
  function VecSub(va, vb) {
    return [va[0] - vb[0], va[1] - vb[1], va[2] - vb[2]];
  }

  // cross product of vectors
  function VecCross(va, vb) {
    return [va[1] * vb[2] - va[2] * vb[1],
            va[2] * vb[0] - va[0] * vb[2],
            va[0] * vb[1] - va[1] * vb[0]];
  }

  // length of a vector
  function VecLength(v) {
    return Math.sqrt(v[0] * v[0] + v[1] * v[1] + v[2] * v[2]);
  }

  // area of a triangle
  function TriangleArea(a, b, c) {
    var ab = VecSub(b, a), ac = VecSub(c, a);
    return .5 * VecLength(VecCross(ab, ac));
  }

  // multiplication of matrices
  function MatMul(ma, mb) {
    return [[ma[0][0] * mb[0][0] + ma[0][1] * mb[1][0] + ma[0][2] * mb[2][0],
             ma[0][0] * mb[0][1] + ma[0][1] * mb[1][1] + ma[0][2] * mb[2][1],
             ma[0][0] * mb[0][2] + ma[0][1] * mb[1][2] + ma[0][2] * mb[2][2]],
            [ma[1][0] * mb[0][0] + ma[1][1] * mb[1][0] + ma[1][2] * mb[2][0],
             ma[1][0] * mb[0][1] + ma[1][1] * mb[1][1] + ma[1][2] * mb[2][1],
             ma[1][0] * mb[0][2] + ma[1][1] * mb[1][2] + ma[1][2] * mb[2][2]],
            [ma[2][0] * mb[0][0] + ma[2][1] * mb[1][0] + ma[2][2] * mb[2][0],
             ma[2][0] * mb[0][1] + ma[2][1] * mb[1][1] + ma[2][2] * mb[2][1],
             ma[2][0] * mb[0][2] + ma[2][1] * mb[1][2] + ma[2][2] * mb[2][2]]];
  }

  // determinant of a matrix
  function MatDet(m) {
    return m[0][0] * (m[1][1] * m[2][2] - m[1][2] * m[2][1]) -
           m[0][1] * (m[1][0] * m[2][2] - m[1][2] * m[2][0]) +
           m[0][2] * (m[1][0] * m[2][1] - m[1][1] * m[2][0]);
  }

  // transpose of a matrix
  function MatTrs(m) {
    return [[m[0][0], m[1][0], m[2][0]],
            [m[0][1], m[1][1], m[2][1]],
            [m[0][2], m[1][2], m[2][2]]];
  }

  // inverse of a matrix
  function MatInv(m) {
    var p00 = m[1][1] * m[2][2] - m[1][2] * m[2][1];
    var p10 = m[1][2] * m[2][0] - m[1][0] * m[2][2];
    var p20 = m[1][0] * m[2][1] - m[1][1] * m[2][0];

    var t = m[0][0] * p00 + m[0][1] * p10 + m[0][2] * p20;
    if (t == 0.) return;

    var t = 1. / t;

    return [[p00 * t,
             (m[0][2] * m[2][1] - m[0][1] * m[2][2]) * t,
             (m[0][1] * m[1][2] - m[0][2] * m[1][1]) * t],
            [p10 * t,
             (m[0][0] * m[2][2] - m[0][2] * m[2][0]) * t,
             (m[0][2] * m[1][0] - m[0][0] * m[1][2]) * t],
            [p20 * t,
             (m[0][1] * m[2][0] - m[0][0] * m[2][1]) * t,
             (m[0][0] * m[1][1] - m[0][1] * m[1][0]) * t]];
  }

  // eigen decomposition of a matrix
  function MatEig(m) {
    if (m[0][1] != m[1][0] ||
        m[0][2] != m[2][0] ||
        m[1][2] != m[2][1])
      return;

    var res = {val : [0., 0., 0.],
               vec : [[1., 0., 0.],
                      [0., 1., 0.],
                      [0., 0., 1.]]};

    var m00 = m[0][0], m11 = m[1][1], m22 = m[2][2],
        m01 = m[0][1], m02 = m[0][2], m12 = m[1][2];

    for (var i = 0; i < kMaxSweeps; ++i) {
      if ((Math.abs(m01) < kEps) && (Math.abs(m02) < kEps) &&
          (Math.abs(m12) < kEps)) break;

      if (m01 != 0.) {
        var u = (m11 - m00) * .5 / m01;
        var u2 = u * u;
        var u2p1 = u2 + 1.;
        var t = (u2p1 != u2) ?
                ((u < 0.) ? -1. : 1.) * (Math.sqrt(u2p1) - Math.abs(u)) :
                .5 / u;
        var c = 1. / Math.sqrt(t * t + 1.);
        var s = c * t;
        m00 -= t * m01;
        m11 += t * m01;
        m01 = 0.;

        var tmp = c * m02 - s * m12;
        m12 = s * m02 + c * m12;
        m02 = tmp;

        for (var j = 0; j < 3; ++j) {
          var tmp = c * res.vec[j][0] - s * res.vec[j][1];
          res.vec[j][1] = s * res.vec[j][0] + c * res.vec[j][1];
          res.vec[j][0] = tmp;
        }
      }

      if (m02 != 0.) {
        var u = (m22 - m00) * .5 / m02;
        var u2 = u * u;
        var u2p1 = u2 + 1.;
        var t = (u2p1 != u2) ?
                ((u < 0.) ? -1. : 1.) * (Math.sqrt(u2p1) - Math.abs(u)) :
                .5 / u;
        var c = 1. / Math.sqrt(t * t + 1.);
        var s = c * t;

        m00 -= t * m02;
        m22 += t * m02;
        m02 = 0.;

        var tmp = c * m01 - s * m12;
        m12 = s * m01 + c * m12;
        m01 = tmp;

        for (var j = 0; j < 3; ++j) {
          var tmp = c * res.vec[j][0] - s * res.vec[j][2];
          res.vec[j][2] = s * res.vec[j][0] + c * res.vec[j][2];
          res.vec[j][0] = tmp;
        }
      }

      if (m12 != 0.) {
        var u = (m22 - m11) * .5 / m12;
        var u2 = u * u;
        var u2p1 = u2 + 1.;
        var t = (u2p1 != u2) ?
                ((u < 0.) ? -1. : 1.) * (Math.sqrt(u2p1) - Math.abs(u)) :
                .5 / u;
        var c = 1. / Math.sqrt(t * t + 1.);
        var s = c * t;

        m11 -= t * m12;
        m22 += t * m12;
        m12 = 0.;

        var tmp = c * m01 - s * m02;
        m02 = s * m01 + c * m02;
        m01 = tmp;

        for (var j = 0; j < 3; ++j) {
          var tmp = c * res.vec[j][1] - s * res.vec[j][2];
          res.vec[j][2] = s * res.vec[j][1] + c * res.vec[j][2];
          res.vec[j][1] = tmp;
        }
      }
    }

    res.val[0] = m00;
    res.val[1] = m11;
    res.val[2] = m22;

    return res;
  }

  // square root of a matrix
  function MatSqrt(m) {
    var eig = MatEig(m);
    if (eig === undefined) return;
    if (eig.val[0] < 0. || eig.val[1] < 0. || eig.val[2] < 0.)
      return;

    var diag = [[Math.sqrt(eig.val[0]), 0., 0.],
                [0., Math.sqrt(eig.val[1]), 0.],
                [0., 0., Math.sqrt(eig.val[2])]];

    return MatMul(MatMul(eig.vec, diag), MatTrs(eig.vec));
  }
  // ---------------------------------------------------------------------------
}
